src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1890 @@
+    # Synchronize Colors #
+    ######################
+
+    def _sync_colors(self, Rijs):
+        """
+        At this point, we have obtained a hand-consistent set of 4-tuples of Rijs,
+        with Rij = Ri.T @ g_s @ Rj, where s is an unknown permutation of (0, 1, 2, 3).
+        Taking the average of the first Rij in the 4-tuple with the remaining 3
+        results in a set of 3 outer products (+-)vi(m).T @ vj(m) of unknown ordering
+        where vi(m) is the m'th row of the i'th rotation matrix, Ri.
+
+        The color sync procedure partitions the set of 3-tuples of m'th row outer
+        products into 3 sets of row-consistent outer products up to the sign of each.
+        """
+        logger.info("Performing rotations' rows synchronization.")
+        # Generate array of one rank matrices from which we can extract rows.
+        # Matrices are of the form 0.5(Ri^TRj+Ri^TgkRj). Each such matrix can be
+        # written in the form Qi^T*Ik*Qj where Ik is a 3x3 matrix with all zero
+        # entries except for the entry a_kk, k in {1,2,3}.
+        n_pairs = len(Rijs)
+        Rijs_rows = np.zeros((n_pairs, 3, 3, 3), dtype=self.dtype)
+        for layer in range(3):
+            Rijs_rows[:, layer] = 0.5 * (Rijs[:, 0] + Rijs[:, layer + 1])
+
+        # Partition the set of matrices Rijs_rows into 3 sets of matrices, where
+        # each set there are only matrices Qi^T*Ik*Qj for a unique value of k in
+        # {1,2,3}.
+        # First determine for each pair of tuples of the form {Qi^T*Ik*Qj} and
+        # {Qr^T*Il*Qj} where {i,j}\cap{r,l}==1, whether l==r.
+        color_perms = self._match_colors(Rijs_rows)
+
+        # Compute eigenvectors of color matrix. This is just a matrix of dimensions
+        # 3(N choose 2)x3(N choose 2) where each entry corresponds to a pair of
+        # matrices {Qi^T*Ir*Qj} and {Qr^T*Il*Qj} and equals \delta_rl.
+        # The 2 leading eigenvectors span a linear subspace which contains a
+        # vector which encodes the partition. All the entries of the vector are
+        # either 1,0 or -1, where the number encodes which the index r in Ir.
+        # This vector is a linear combination of the two leading eigen vectors,
+        # and so we 'unmix' these vectors to retrieve it.
+        color_mat = la.LinearOperator(
+            (3 * n_pairs,) * 2, lambda v: self._mult_cmat_by_vec(color_perms, v)
+        )
+        v0 = randn(
+            3 * n_pairs, seed=self.seed
+        )  # Seed eigs initial vector for iterative method
+        v0 = v0 / norm(v0)
+        vals, colors = la.eigs(color_mat, k=3, which="LR", v0=v0)
+        vals = np.real(vals)
+        colors = np.real(colors)
+        colors = np.sign(colors[0]) * colors  # Stable eigs
+        cp, _ = self._unmix_colors(colors[:, :2])
+
+        return cp, Rijs_rows
+
+    def _match_colors(self, Rijs_rows):
+        """
+        Partition the set of matrices Rijs_rows, which correspond to a permutation of
```

Comment:
I don't see how the return value performs a partition here.

Reply:
You are correct. I rewrote this description to indicate that we are finding permutations for each triplet of `Rijs_rows` (ie. the set of m'th row outer products of unknown order) that reorders the elements of each `Rijs_rows` so that the triplets all have the same ordering. (Sorry if that sounds confusing).

At any rate, hopefully the description is better describing what is happening. I also added an explanation of the term "color matching".

Reply:
Cool. Thanks!